relay-cardinality/src/redis/state.rs

```diff
@@ -0,0 +1,175 @@
+use std::collections::BTreeMap;
+
+use relay_statsd::metric;
+
+use crate::{
+    limiter::{Entry, EntryId, Scoping},
+    redis::quota::{PartialQuotaScoping, QuotaScoping},
+    statsd::{CardinalityLimiterCounters, CardinalityLimiterSets},
+    CardinalityLimit,
+};
+
+/// Internal state combining relevant entries for the respective quotas.
+///
+/// Also tracks amount of cache hits and misses as well as amount of
+/// items accepted and rejected. These metrics are reported via statsd on drop.
+pub struct LimitState<'a> {
+    /// The limit of the quota.
+    pub limit: u64,
+
+    /// Scoping of the quota.
+    scope: PartialQuotaScoping,
+    /// Entries grouped by quota scoping.
+    ///
+    /// Contained scopes are a superset of `scope`.
+    sub_scopes: BTreeMap<QuotaScoping, Vec<RedisEntry>>,
```

Comment:
Similar question as above: Why is there a single limit state holding multiple `name`s , instead of a separate limit state per name?

Reply:
There is a 1:1 relationship between cardinality limit and a limit state, which helps with a few things:

- Everything that is related to the cardinality limit is tracked in a single state (e.g. all metrics).
- There is known fixed amount of states independently of the amount of items passed in, which means there is no need to have 2 phases of "gather all states by looping through the items" and a 2nd phase of "adding items to their states".
- We can use the already setup `PartialQuotaScoping` to "filter" the metrics down and only then branch out and create more scopes when necessary.
- Additionally the `PartialQuotaScope` defines the Redis sharding key, having the "full" scopes already grouped by their common partial scopes makes batching very easy instead of having another phase which groups all the independent states together.